test/Microsoft.AspNetCore.Razor.Test.Common/Language/IntegrationTests/IntegrationTestBase.cs

```diff
@@ -220,9 +220,102 @@ protected void AssertSourceMappingsMatchBaseline(RazorCodeDocument document)
             var baseline = testFile.ReadAllText();
 
             // Normalize newlines to match those in the baseline.
-            var actual = serializedMappings.Replace("\r", "").Replace("\n", "\r\n");
+            var actualBaseline = serializedMappings.Replace("\r", "").Replace("\n", "\r\n");
 
-            Assert.Equal(baseline, actual);
+            Assert.Equal(baseline, actualBaseline);
+
+            var syntaxTree = codeDocument.GetSyntaxTree();
+            var visitor = new CodeSpanVisitor();
+            visitor.VisitBlock(syntaxTree.Root);
+
+            var charBuffer = new char[codeDocument.Source.Length];
+            codeDocument.Source.CopyTo(0, charBuffer, 0, codeDocument.Source.Length);
+            var sourceContent = new string(charBuffer);
+
+            var spans = visitor.CodeSpans;
+            for (var i= 0; i < spans.Count; i++)
+            {
+                var span = spans[i];
+                if (span.Start.FilePath == null || span.Start.FilePath != codeDocument.Source.FilePath)
+                {
+                    // Not in the main file, skip.
+                    continue;
+                }
+
+                var location = new SourceSpan(span.Start, span.Length);
+                var expectedSpan = sourceContent.Substring(span.Start.AbsoluteIndex, span.Length);
+
+                // See #
+                if (string.IsNullOrWhiteSpace(expectedSpan))
+                {
+                    // For now we don't verify whitespace inside of a directive. We know that directives cheat
+                    // with how they bound whitespace/C#/markup to make completion work.
+                    if (span.Parent is Block block && block.Type == BlockKindInternal.Directive)
+                    {
+                        continue;
+                    }
+                }
+
+                // See #
```

Comment:
I'm planning to log a bug for this, I expect it will end up in the same place as the above. We categorize the second `@` in `@{ @@Da }` as C# but then we don't put it in the buffer. I expect that this is done for similar reasons, and that changing it could break completion.